Content.Server/GameObjects/Components/Kitchen/ReagentGrinderComponent.cs

```diff
@@ -0,0 +1,367 @@
+        {
+            base.ExposeData(serializer);    
+            serializer.DataField(ref _storageCap, "chamber_capacity", 16);
+            serializer.DataField(ref _workTime, "time", 3500);
+        }
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            //A slot for the beaker where the grounds/juices will go.
+            _beakerContainer =
+                ContainerManagerComponent.Ensure<ContainerSlot>($"{Name}-reagentContainerContainer", Owner);
+
+            //A container for the things that WILL be ground/juiced. Useful for ejecting them instead of deleting them from the hands of the user.
+            _chamber =
+                ContainerManagerComponent.Ensure<Container>($"{Name}-entityContainerContainer", Owner);
+
+            if (UserInterface != null)
+            {
+                UserInterface.OnReceiveMessage += UserInterfaceOnReceiveMessage;
+            }
+
+            _audioSystem = EntitySystem.Get<AudioSystem>();
+        }
+
+        private void UserInterfaceOnReceiveMessage(ServerBoundUserInterfaceMessage message)
+        {
+            if(!Powered || _busy)
+            {
+                return;
+            }
+
+            switch(message.Message)
+            {
+                case ReagentGrinderGrindStartMessage msg:
+                    ClickSound();
+                    DoWork(user:message.Session.AttachedEntity!,isJuiceIntent:false);
+                    break;
+
+                case ReagentGrinderJuiceStartMessage msg:
+                    ClickSound();
+                    DoWork(user: message.Session.AttachedEntity!,isJuiceIntent:true);
+                    break;
+
+                case ReagentGrinderEjectChamberAllMessage msg:
+                    if(!ChamberEmpty)
+                    {
+                        ClickSound();
+                        for (var i = _chamber.ContainedEntities.Count - 1; i >= 0; i--)
+                        {
+                            EjectSolid(_chamber.ContainedEntities.ElementAt(i).Uid);
+                        }
+                    }
+                    break;
+
+                case ReagentGrinderEjectChamberContentMessage msg:
+                    if (!ChamberEmpty)
+                    {
+                        EjectSolid(msg.EntityID);
+                        ClickSound();
+                        _dirty = true;
+                    }
+                    break;
+
+                case ReagentGrinderEjectBeakerMessage msg:
+                    ClickSound();
+                    EjectBeaker(message.Session.AttachedEntity!);
+                    //EjectBeaker will dirty the UI for us, we don't have to do it explicitly here.
+                    break;
+            }
+        }
+        private void ClickSound()
+        {
+            _audioSystem.PlayFromEntity("/Audio/Machines/machine_switch.ogg", Owner, AudioParams.Default.WithVolume(-2f));
```

Comment:
IMO just resolve it when you need it given it's rarely pressed and it'll save a bit of mem.

Reply:
its gonnna get called very often but if its more mem efficient that way then thats even better.